block/bench/Block.hs

```diff
@@ -0,0 +1,249 @@
+    , tsSscPayload  :: !(SscPayload, LBS.ByteString)
+    , tsDlgPayload  :: !(DlgPayload, LBS.ByteString)
+    , tsUpdPayload  :: !(UpdatePayload, LBS.ByteString)
+    , tsExtraBody   :: !(ExtraBodyData MainBlockchain, LBS.ByteString)
+    }
+
+instance NFData TestSubject where
+    rnf (TestSubject {..}) =
+        deepseq tsBlock $
+          deepseq tsHeader $
+          deepseq tsBodyProof $
+          deepseq tsConsensus $
+          deepseq tsExtraHeader $
+          deepseq tsBody $
+          deepseq tsTxPayload $
+          deepseq tsSscPayload $
+          deepseq tsDlgPayload $
+          deepseq tsUpdPayload $
+          deepseq tsExtraBody $
+          ()
+
+testSubjectSize :: (a, LBS.ByteString) -> Int64
+testSubjectSize (_, bytes) = LBS.length bytes
+
+printSize :: TestSubject -> String -> (TestSubject -> (a, LBS.ByteString)) -> IO ()
+printSize ts name f = Universum.putStrLn $ name <> " of size " <> Universum.show (testSubjectSize (f ts)) <> " bytes"
+
+printSizes :: TestSubject -> IO TestSubject
+printSizes ts = do
+    printSize ts "block" tsBlock
+    printSize ts "header" tsHeader
+    printSize ts "body proof" tsBodyProof
+    printSize ts "consensus" tsConsensus
+    printSize ts "extra header data" tsExtraHeader
+    printSize ts "body" tsBody
+    printSize ts "tx payload" tsTxPayload
+    printSize ts "ssc payload" tsSscPayload
+    printSize ts "dlg payload" tsDlgPayload
+    printSize ts "upd payload" tsUpdPayload
+    printSize ts "extra body data" tsExtraBody
+    return ts
+
+withSerialized :: Bi a => a -> (a, LBS.ByteString)
+withSerialized a = (a, serialize a)
+
+-- | Make a TestSubject using a seed for a PRNG and size.
+testSubject
+    :: ( HasConfiguration )
+    => Int -- ^ Seed
+    -> Int -- ^ Size
+    -> TestSubject
+testSubject seed size =
+  let block :: MainBlock
+      block = generateBlock seed size
+
+      tsBlock = withSerialized block
+      tsHeader = withSerialized (_gbHeader $ block)
+      tsBodyProof = withSerialized (_gbhBodyProof . _gbHeader $ block)
+      tsConsensus = withSerialized (_gbhConsensus . _gbHeader $ block)
+      tsExtraHeader = withSerialized (_gbhExtra . _gbHeader $ block)
+      tsBody = withSerialized (_gbBody $ block)
+      tsTxPayload = withSerialized (_mbTxPayload . _gbBody $ block)
+      tsSscPayload = withSerialized (_mbSscPayload . _gbBody $ block)
+      tsDlgPayload = withSerialized (_mbDlgPayload . _gbBody $ block)
+      tsUpdPayload = withSerialized (_mbUpdatePayload . _gbBody $ block)
+      tsExtraBody  = withSerialized (_gbExtra $ block)
+
+  in  TestSubject {..}
+
+benchMain :: ( HasConfiguration ) => Int -> Int -> IO ()
+benchMain seed size = defaultMain
+    [ env (return (testSubject seed size) >>= printSizes) $ \ts -> bgroup "block" $
+          [ bgroup "serialize" $ 
+                [ bench "all" (nf serialize (fst . tsBlock $ ts))
+                , bgroup "header" $
+                      [ bench "all" (nf serialize (fst . tsHeader $ ts))
```

Comment:
Why not `whatever t f = bench t (nf serialize (fst $ f ts))` and use `whatever` instead? Same goes for the `bgroup "deserialize"` copy-paste.

Reply:
Sure. But then it'll be `whatever "deserialize"` copy-paste so what do we really gain?

Reply:
Less copy-paste, which is a default strategy of eliminating copy-paste :p